### PR TITLE
PCHR-2152: Remind Me Functionality

### PR DESCRIPTION
**Pull request by mickadoo**
## Overview
The functionality for cloning documents that are approaching their expiry date is already developed and in use. After adding the "remind me" column in #189 we need to check before cloning if the user has selected that they want to be reminded about the document before cloning it.

## Before
All documents that had an expiry date soon approaching were cloned.

## After
Only documents with "remind me" set to true will be cloned (if their expiry date is approaching).

## Technical Details
- Originally I had to unset all the custom field names (there are 3 variations for each e.g. `expire_date`, `custom_77` & `custom_77_9`) before creating but by using "return" in the get request it limits the field names.

## Comments
- I tried to tidy up by removing raw SQL usage

---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 1]
```diff
@@ -11,7 +11,7 @@ class CRM_Tasksassignments_BAO_Document extends CRM_Tasksassignments_DAO_Documen
    * Create a new Document based on array-data
    *
    * @param array $params key-value pairs
-   * @return CRM_Tasksassignments_DAO_Document|NULL
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: Why did you need to add the object type? Note that we already say this method returns a Document DAO instance

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 2]
```diff
@@ -38,118 +38,115 @@ public static function create(&$params) {
...
 
   /**
-   * Return a 'days_to_create_a_document_clone' value from T&A Settings.
-   * 
-   * @return int
+   * Finds all documents that should be cloned
+   *
+   * @param DateTime $cutOffDate
+   *  Documents with an expiry date less than this will be cloned.
+   * @return array
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: Please add an empty line between the `@param` and `@return` sections

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 3]
```diff
@@ -38,118 +38,115 @@ public static function create(&$params) {
...
   }
 
   /**
-   * Return a 'days_to_create_a_document_clone' value from T&A Settings.
-   * 
-   * @return int
+   * Finds all documents that should be cloned
+   *
+   * @param DateTime $cutOffDate
+   *  Documents with an expiry date less than this will be cloned.
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: Here you say less than, but on the API call you use `<=`. Which one is correct?

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 4]
```diff
@@ -38,118 +38,115 @@ public static function create(&$params) {
...
+    $remindMeField = static::getCustomFieldName('remind_me');
+
+    $params = [
+      $expiryField => ['<=' => $cutOffDate->format('Y-m-d')],
+      $clonedField => ['IS NULL' => 1],
+      $remindMeField => 1,
+      'is_deleted' => 0,
+      'status_id' => self::STATUS_APPROVED,
+      'options' => ['limit' => 0],
+      'return' => ["target_contact_id", "activity_type_id", "details"]
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: please use single quotes for the field names

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 5]
```diff
@@ -38,118 +38,115 @@ public static function create(&$params) {
...
+    $fieldsToUnset = ['id', 'file_count', 'expire_date', 'document_number', 'valid_from'];
+    foreach ($fieldsToUnset as $field) {
+      unset($clone[$field]);
     }
-    return $count;
+
+    civicrm_api3('Document', 'create', $clone);
+
+    // Update original's clone_date' so we know it has been cloned.
+    $today = (new DateTime())->format('Y-m-d');
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: I like using `DateTime` objects, but here I think a simple `date('Y-m-d')` would be better and more readable. What do you think?

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 6]
```diff
@@ -38,118 +38,115 @@ public static function create(&$params) {
...
   /**
-   * Search for approved Documents having expire date less than given $expireDate
-   * and create their clone.
-   * Return a count of cloned documents.
-   * 
-   * @param DateTime $expireDate
-   * @return int
+   * Creates a copy of a document with no attached files
+   *
+   * @param array $original
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: Please add a description to this param

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 7]
```diff
@@ -38,118 +38,115 @@ public static function create(&$params) {
...
-   * PCHR-1365 requirements.
-   * 
-   * @param array $document
-   * @param int $expireId
-   * @return array
+   * Gets the custom field name for Documents
+   *
+   * @param string $fieldName
+   *  The human-readable name for the field, e.g. "expiry_date"
+   * @return mixed|null
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: Please add an empty line between the param and return sections

Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: Why mixed? Shouldn't it be string|null?

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php (hunk 1 above)
Comment (mickadoo) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: I can delete it, just the IDE doesn't know that $instance is a CRM_Tasksassignments_DAO_Document so I added it to hide the warning that it was returning something unexpected. The parent class also has a return which includes `object`.

image

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php (hunk 3 above)
Comment (mickadoo) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: I'll add "or equal to"

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php (hunk 5 above)
Comment (mickadoo) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: I've found working with DateTime much more readable in the majority of cases, particularly when adding / subtracting intervals but I think in this case there's no need. Also I guess just because they introduced `DateTime` doesn't mean they're planning to deprecate other date functions.

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php (hunk 7 above)
Comment (mickadoo) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: It's just expecting the same return value as `CRM_Utils_Array::value`, which is `mixed`, it was autogenerated and since it doesn't complain if I use `string` I'll change it.

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php [hunk 8]
```diff
@@ -38,118 +39,116 @@ public static function create(&$params) {
...
-      // so so we'll know that the document has been already cloned.
-      civicrm_api3('Document', 'create', array(
-        'id' => $originalId,
-        'clone_date' => (new DateTime())->format('Y-m-d'),
-      ));
+  private static function getCustomFieldName($fieldName) {
+    static $fieldNames;
+
+    if (empty($fieldNames)) {
+      $result = civicrm_api3('Document', 'getcustomfields');
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php]: @davialexandre I also made changes here since your review. I think this is simpler than those fancy array functions

**Comment by davialexandre** [comment]
Approving the changes even though the tests are not passing, so it doesn't block further development on PCHR-2145. @mickadoo will work on fixing the tests and will submit a separate PR for this

File: uk.co.compucorp.civicrm.tasksassignments/tests/phpunit/api/v3/DocumentTest.php [hunk 9]
```diff
@@ -170,4 +171,65 @@ function testCreateDocumentClone() {
...
+
+    $contact = civicrm_api3('Contact', 'create', [
+      'contact_type' => 'Individual',
+      'display_name' => 'First Contact',
+      'email' => 'test1@email3215155.com',
+    ]);
+    $details1 = 'First sample document details';
+    $details2 = 'Second sample document details';
+    $today = (new DateTime())->format('Y-m-d');
+    $tomorrow = (new DateTime('tomorrow'))->format('Y-m-d');
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/tests/phpunit/api/v3/DocumentTest.php]: if you're not really using the `DateTime` instances, I think this will be simpler with the `date()` function

File: uk.co.compucorp.civicrm.tasksassignments/tests/phpunit/api/v3/TaskTest.php [hunk 10]
```diff
@@ -1,69 +1,75 @@
...
     $upgrader->install();
 
-    // Pick one of Task types.
-    $taskTypes = civicrm_api3('Task', 'getoptions', array(
-      'field' => 'activity_type_id',
-    ));
-    $this->_taskTypeId = array_shift($taskTypes['values']);
-  }
+    $result = civicrm_api3('OptionValue', 'create', [
+      'option_group_id' => "activity_type",
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/tests/phpunit/api/v3/TaskTest.php]: use single quotes for everything here

File: uk.co.compucorp.civicrm.tasksassignments/tests/phpunit/api/v3/TaskTest.php [hunk 11]
```diff
@@ -1,69 +1,75 @@
...
     $upgrader = CRM_Tasksassignments_Upgrader::instance();
     $upgrader->install();
 
-    // Pick one of Task types.
-    $taskTypes = civicrm_api3('Task', 'getoptions', array(
-      'field' => 'activity_type_id',
-    ));
-    $this->_taskTypeId = array_shift($taskTypes['values']);
-  }
+    $result = civicrm_api3('OptionValue', 'create', [
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/tests/phpunit/api/v3/TaskTest.php]: Unless you have a good reason not to, you should use the `tearDown` method to deleted anything you create on a `setUp` method.

File: uk.co.compucorp.civicrm.tasksassignments/api/v3/Document.php [hunk 12]
```diff
@@ -242,6 +242,8 @@ function _civicrm_api3_document_create_spec(&$params) {
       'api.default' => 'user_contact_id',
   ];
 
+  $params['id']['title'] = "Document ID";
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/api/v3/Document.php]: single quotes
